File: Source/TitaniumKit/src/Gesture.cpp
/**
 * TitaniumKit
 *
 * Copyright (c) 2014-2015 by Appcelerator, Inc. All Rights Reserved.
 * Licensed under the terms of the Apache Public License.
 * Please see the LICENSE included with this distribution for details.
 */

#include "Titanium/Gesture.hpp"
#include "Titanium/detail/TiImpl.hpp"
#include <type_traits>

namespace Titanium
{
	Gesture::Gesture(const JSContext& js_context) TITANIUM_NOEXCEPT
	    : Module(js_context, "Titanium.Gesture")
	{
	}

	Titanium::UI::ORIENTATION Gesture::get_orientation() const TITANIUM_NOEXCEPT
	{
		TITANIUM_LOG_WARN("Gesture::get_orientation: Unimplemented");
		return Titanium::UI::ORIENTATION::UNKNOWN;
	}

	void Gesture::JSExportInitialize()
	{
		JSExport<Gesture>::SetClassVersion(1);
		JSExport<Gesture>::SetParent(JSExport<Module>::Class());
		
		TITANIUM_ADD_FUNCTION(Gesture, getLandscape);
		TITANIUM_ADD_FUNCTION(Gesture, getPortrait);
		TITANIUM_ADD_FUNCTION(Gesture, getOrientation);
		TITANIUM_ADD_FUNCTION(Gesture, isFaceDown);
		TITANIUM_ADD_FUNCTION(Gesture, isFaceUp);
		TITANIUM_ADD_FUNCTION(Gesture, isLandscape);
		TITANIUM_ADD_FUNCTION(Gesture, isPortrait);
		TITANIUM_ADD_PROPERTY_READONLY(Gesture, orientation);
		TITANIUM_ADD_PROPERTY_READONLY(Gesture, landscape);
		TITANIUM_ADD_PROPERTY_READONLY(Gesture, portrait);
	}

	JSObject Gesture::GetStaticObject(const JSContext& js_context) TITANIUM_NOEXCEPT
	{
		JSValue Titanium_property = js_context.get_global_object().GetProperty("Titanium");
		TITANIUM_ASSERT(Titanium_property.IsObject());  // precondition
		JSObject Titanium = static_cast<JSObject>(Titanium_property);

		JSValue Object_property = Titanium.GetProperty("Gesture");
		TITANIUM_ASSERT(Object_property.IsObject());  // precondition
		return static_cast<JSObject>(Object_property);
	}

	TITANIUM_PROPERTY_GETTER(Gesture, orientation)
	{
		return get_context().CreateNumber(UI::Constants::to_underlying_type(get_orientation()));
	}
	
	TITANIUM_FUNCTION(Gesture, getOrientation)
	{
		const auto js_context = this_object.get_context();
		const auto gesture_ptr = GetStaticObject(js_context).GetPrivate<Gesture>();
		return this_object.get_context().CreateNumber(UI::Constants::to_underlying_type(gesture_ptr->get_orientation()));
	}

	TITANIUM_FUNCTION(Gesture, isFaceDown)
	{
		const auto js_context = this_object.get_context();
		const auto gesture_ptr = GetStaticObject(js_context).GetPrivate<Gesture>();
		return this_object.get_context().CreateBoolean(gesture_ptr->get_orientation() == UI::ORIENTATION::FACE_DOWN);
	}

	TITANIUM_FUNCTION(Gesture, isFaceUp)
	{
		const auto js_context = this_object.get_context();
		const auto gesture_ptr = GetStaticObject(js_context).GetPrivate<Gesture>();
		return this_object.get_context().CreateBoolean(gesture_ptr->get_orientation() == UI::ORIENTATION::FACE_UP);
	}

	TITANIUM_FUNCTION(Gesture, isLandscape)
	{
		const auto js_context = this_object.get_context();
		const auto gesture_ptr = GetStaticObject(js_context).GetPrivate<Gesture>();
		return gesture_ptr->js_get_landscape();
	}

	TITANIUM_FUNCTION(Gesture, isPortrait)
	{
		const auto js_context = this_object.get_context();
		const auto gesture_ptr = GetStaticObject(js_context).GetPrivate<Gesture>();
		return gesture_ptr->js_get_portrait();
	}

	TITANIUM_FUNCTION(Gesture, getLandscape)
	{
		const auto js_context = this_object.get_context();
		const auto gesture_ptr = GetStaticObject(js_context).GetPrivate<Gesture>();
		return gesture_ptr->js_get_landscape();
	}

	TITANIUM_PROPERTY_GETTER(Gesture, landscape)
	{
		const auto o = get_orientation();
		const bool isLandscape = (o == UI::ORIENTATION::LANDSCAPE_LEFT || o == UI::ORIENTATION::LANDSCAPE_RIGHT);
		return get_context().CreateBoolean(isLandscape);
	}

	TITANIUM_FUNCTION(Gesture, getPortrait)
	{
		const auto js_context = this_object.get_context();
		const auto gesture_ptr = GetStaticObject(js_context).GetPrivate<Gesture>();
		return gesture_ptr->js_get_portrait();
	}

	TITANIUM_PROPERTY_GETTER(Gesture, portrait)
	{
		const auto o = get_orientation();
		const bool isPortrait = (o == UI::ORIENTATION::PORTRAIT || o == UI::ORIENTATION::UPSIDE_PORTRAIT);
		return get_context().CreateBoolean(isPortrait);
	};

	void Gesture::enableEvent(const std::string& event_name) TITANIUM_NOEXCEPT
	{
		TITANIUM_LOG_WARN("Ti::Gesture::enableEvent: Unimplemented (event name '", event_name, "'");
	}

	void Gesture::disableEvent(const std::string& event_name) TITANIUM_NOEXCEPT
	{
		TITANIUM_LOG_WARN("Ti::Gesture::disableEvent: Unimplemented (event name '", event_name, "'");
	}

}  // namespace Titanium

File: Examples/Mocha/src/Assets/gesture.js
/*
 * Appcelerator Titanium Mobile
 * Copyright (c) 2011-2014 by Appcelerator, Inc. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */

require('ti-mocha');
var should = require('should');

describe('gesture', function () {
    it('Ti.Gesture', function (finish) {
        should(Ti.Gesture).not.be.undefined;
        should(Ti.Gesture).be.an.Object;
        should(Ti.Gesture.addEventListener).be.a.Function;
        should(Ti.Gesture.removeEventListener).be.a.Function;
        finish();
    });
    it('Ti.Gesture.landscape', function (finish) {
        should(Ti.Gesture.landscape).not.be.undefined;
        should(Ti.Gesture.landscape).be.a.Boolean;
        finish();
    });
    it('Ti.Gesture.portrait', function (finish) {
        should(Ti.Gesture.portrait).not.be.undefined;
        should(Ti.Gesture.portrait).be.a.Boolean;
        finish();
    });
    it('Ti.Gesture.orientation', function (finish) {
        should(Ti.Gesture.orientation).not.be.undefined;
        should(Ti.Gesture.orientation).be.a.Number;
        finish();
    });
    it('Ti.UI.PORTRAIT', function (finish) {
        should(Ti.UI.PORTRAIT).not.be.undefined;
        should(Ti.UI.PORTRAIT).be.a.Number;
        finish();
    });
    it('Ti.UI.UPSIDE_PORTRAIT', function (finish) {
        should(Ti.UI.UPSIDE_PORTRAIT).not.be.undefined;
        should(Ti.UI.UPSIDE_PORTRAIT).be.a.Number;
        finish();
    });
    it('Ti.UI.LANDSCAPE_LEFT', function (finish) {
        should(Ti.UI.LANDSCAPE_LEFT).not.be.undefined;
        should(Ti.UI.LANDSCAPE_LEFT).be.a.Number;
        finish();
    });
    it('Ti.UI.LANDSCAPE_RIGHT', function (finish) {
        should(Ti.UI.LANDSCAPE_RIGHT).not.be.undefined;
        should(Ti.UI.LANDSCAPE_RIGHT).be.a.Number;
        finish();
    });
    it('Ti.UI.FACE_UP', function (finish) {
        should(Ti.UI.FACE_UP).not.be.undefined;
        should(Ti.UI.FACE_UP).be.a.Number;
        finish();
    });
    it('Ti.UI.UNKNOWN', function (finish) {
        should(Ti.UI.UNKNOWN).not.be.undefined;
        should(Ti.UI.UNKNOWN).be.a.Number;
        finish();
    });
    it('Ti.Gesture.getLandscape', function (finish) {
        should(Ti.Gesture.getLandscape).not.be.undefined;
        should(Ti.Gesture.getLandscape).be.a.Function;
        should(Ti.Gesture.getLandscape()).be.a.Boolean;
        finish();
    });
    it('Ti.Gesture.getPortrait', function (finish) {
        should(Ti.Gesture.getPortrait).not.be.undefined;
        should(Ti.Gesture.getPortrait).be.a.Function;
        should(Ti.Gesture.getPortrait()).be.a.Boolean;
        finish();
    });
    it('Ti.Gesture.isLandscape', function (finish) {
        should(Ti.Gesture.isLandscape).not.be.undefined;
        should(Ti.Gesture.isLandscape).be.a.Function;
        should(Ti.Gesture.isLandscape()).be.a.Boolean;
        finish();
    });
    it('Ti.Gesture.isPortrait', function (finish) {
        should(Ti.Gesture.isPortrait).not.be.undefined;
        should(Ti.Gesture.isPortrait).be.a.Function;
        should(Ti.Gesture.isPortrait()).be.a.Boolean;
        finish();
    });
    it('Ti.Gesture.isFaceDown', function (finish) {
        should(Ti.Gesture.isFaceDown).not.be.undefined;
        should(Ti.Gesture.isFaceDown).be.a.Function;
        should(Ti.Gesture.isFaceDown()).be.a.Boolean;
        finish();
    });
    it('Ti.Gesture.isFaceUp', function (finish) {
        should(Ti.Gesture.isFaceUp).not.be.undefined;
        should(Ti.Gesture.isFaceUp).be.a.Function;
        should(Ti.Gesture.isFaceUp()).be.a.Boolean;
        finish();
    });
    it('Ti.Gesture.getOrientation', function (finish) {
        should(Ti.Gesture.getOrientation).not.be.undefined;
        should(Ti.Gesture.getOrientation).be.a.Function;
        should(Ti.Gesture.getOrientation()).be.a.Number;
        finish();
    });
});
